util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,226 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <utility>
+
+#include "channel.h"
+
+namespace concord::channel {
+
+// Forward declarations
+template <typename Msg>
+class BoundedMpscSender;
+template <typename Msg>
+class BoundedMpscReceiver;
+
+namespace impl {
+
+// A simple, non-optimized, lock based, bounded MPSC channel.
+//
+// This is an internal type used by the BoundedMpscSender and BoundedMpscReceiver and should not be exposed publicly.
+template <typename Msg>
+class BoundedMpscQueue {
+ public:
+  BoundedMpscQueue(size_t capacity) : capacity_(capacity) {}
+
+ private:
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscSender;
+  template <typename T>
+  friend class ::concord::channel::BoundedMpscReceiver;
+
+  std::optional<size_t> size() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    return buf_.size();
+  }
+
+  std::optional<size_t> capacity() {
+    // No need for a lock, as capacity_ is immutable
+    return capacity_;
+  }
+
+  void addSender() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    num_senders_++;
+  }
+
+  void removeSender() {
+    bool num_senders = 0;
```

Comment:
I would assume this should be of type `size_t`.

Reply:
wow, good catch.

Reply:
{}-init syntax :)